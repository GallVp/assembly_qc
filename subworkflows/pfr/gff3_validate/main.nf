include { GT_GFF3                               } from '../../../modules/pfr/gt/gff3/main'
include { GT_GFF3VALIDATOR                      } from '../../../modules/pfr/gt/gff3validator/main'
include { CUSTOM_CHECKGFF3FASTACORRESPONDENCE   } from '../../../modules/pfr/custom/checkgff3fastacorrespondence/main'

workflow GFF3_VALIDATE {

    take:
    ch_gff3     // channel: [ val(meta), gff3 ]
    ch_fasta    // channel: [ val(meta), fasta ]

    main:

    ch_versions = Channel.empty()

    // MODULE: GT_GFF3
    GT_GFF3 ( ch_gff3 )
    ch_versions = ch_versions.mix(GT_GFF3.out.versions.first())

    // MODULE: GT_GFF3VALIDATOR
    GT_GFF3VALIDATOR ( GT_GFF3.out.gt_gff3 )
    ch_versions = ch_versions.mix(GT_GFF3VALIDATOR.out.versions.first())

    // MODULE: CUSTOM_CHECKGFF3FASTACORRESPONDENCE
    GT_GFF3VALIDATOR.out.success_log
    | join (
        GT_GFF3.out.gt_gff3
    )
    | map { meta, log, gff3 -> [ meta, gff3 ] }
    | join (
        ch_fasta
    )
    | set { ch_gff3_fasta }

    CUSTOM_CHECKGFF3FASTACORRESPONDENCE (
        ch_gff3_fasta.map { meta, gff3, fasta -> [ meta, gff3 ] },
        ch_gff3_fasta.map { meta, gff3, fasta -> fasta }
    )

    ch_versions = ch_versions.mix(CUSTOM_CHECKGFF3FASTACORRESPONDENCE.out.versions.first())

    CUSTOM_CHECKGFF3FASTACORRESPONDENCE.out.success_log
    | join (
        ch_gff3_fasta.map { meta, gff3, fasta -> [ meta, gff3 ] }
    )
    | map { meta, log, gff3 -> [ meta, gff3 ] }
    | set { ch_valid_gff3 }

    GT_GFF3.out.error_log
    | mix (
        GT_GFF3VALIDATOR.out.error_log
    )
    | mix (
        CUSTOM_CHECKGFF3FASTACORRESPONDENCE.out.error_log
    )
    | set { ch_log_for_invalid_gff3 }

    emit:
    valid_gff3              = ch_valid_gff3             // channel: [ val(meta), gff3 ]
    log_for_invalid_gff3    = ch_log_for_invalid_gff3   // channel: [ val(meta), log ]
    versions                = ch_versions               // channel: [ versions.yml ]
}
