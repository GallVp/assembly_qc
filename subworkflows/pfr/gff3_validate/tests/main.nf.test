nextflow_workflow {

    name "Test Workflow GFF3_VALIDATE"
    script "../main.nf"
    workflow "GFF3_VALIDATE"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/gff3_validate"
    tag "gff3_validate"
    tag "gt"
    tag "gt/gff3"
    tag "gt/gff3validator"
    tag "custom"
    tag "custom/checkgff3fastacorrespondence"

    test("sarscov2-genome_gff3-genome_fasta-all_pass") {

        when {
            workflow {
                """
                input[0] = Channel.of([ [ id:'test' ], // meta map
                    file(params.test_data['sarscov2']['genome']['genome_gff3'], checkIfExists: true)
                ])
                input[1] = Channel.of([ [ id:'test' ],
                    file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("homo_sapiens-genome_bed-genome_fasta-gt_gff3_fail") {

        when {
            workflow {
                """
                input[0] = Channel.of([ [ id:'test' ], // meta map
                    file(params.test_data['homo_sapiens']['genome']['genome_bed'], checkIfExists: true)
                ])
                input[1] = Channel.of([ [ id:'test' ],
                    file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("sarscov2-genome_gff3-homo_sapiens-genome_fasta-correspondence_fail") {

        when {
            workflow {
                """
                input[0] = Channel.of([ [ id:'test' ], // meta map
                    file(params.test_data['sarscov2']['genome']['genome_gff3'], checkIfExists: true)
                ])
                input[1] = Channel.of([ [ id:'test' ],
                    file(params.test_data['homo_sapiens']['genome']['genome_fasta'], checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}