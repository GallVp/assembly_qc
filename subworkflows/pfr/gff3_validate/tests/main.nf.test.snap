{
    "sarscov2-genome_gff3-genome_fasta-all_pass": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.gt.gff3:md5,2ae900237ace415557b8735fac088b85"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    "versions.yml:md5,10fe5c201e5fcddb52c3607ab3fdfb34",
                    "versions.yml:md5,856745cef2fff087e50ea4c0ffa3addd",
                    "versions.yml:md5,a89255422a163684b0c80ebdd8ad28ae"
                ],
                "log_for_invalid_gff3": [
                    
                ],
                "valid_gff3": [
                    [
                        {
                            "id": "test"
                        },
                        "test.gt.gff3:md5,2ae900237ace415557b8735fac088b85"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,10fe5c201e5fcddb52c3607ab3fdfb34",
                    "versions.yml:md5,856745cef2fff087e50ea4c0ffa3addd",
                    "versions.yml:md5,a89255422a163684b0c80ebdd8ad28ae"
                ]
            }
        ],
        "timestamp": "2023-12-07T10:33:21.09887"
    },
    "homo_sapiens-genome_bed-genome_fasta-gt_gff3_fail": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.error.log:md5,c096494c3cd02864eb54434c294ba382"
                    ]
                ],
                "2": [
                    "versions.yml:md5,a89255422a163684b0c80ebdd8ad28ae"
                ],
                "log_for_invalid_gff3": [
                    [
                        {
                            "id": "test"
                        },
                        "test.error.log:md5,c096494c3cd02864eb54434c294ba382"
                    ]
                ],
                "valid_gff3": [
                    
                ],
                "versions": [
                    "versions.yml:md5,a89255422a163684b0c80ebdd8ad28ae"
                ]
            }
        ],
        "timestamp": "2023-12-07T10:35:26.549003"
    },
    "sarscov2-genome_gff3-homo_sapiens-genome_fasta-correspondence_fail": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.error.log:md5,67686ea1ef271821f1218a8fe0207e1f"
                    ]
                ],
                "2": [
                    "versions.yml:md5,10fe5c201e5fcddb52c3607ab3fdfb34",
                    "versions.yml:md5,856745cef2fff087e50ea4c0ffa3addd",
                    "versions.yml:md5,a89255422a163684b0c80ebdd8ad28ae"
                ],
                "log_for_invalid_gff3": [
                    [
                        {
                            "id": "test"
                        },
                        "test.error.log:md5,67686ea1ef271821f1218a8fe0207e1f"
                    ]
                ],
                "valid_gff3": [
                    
                ],
                "versions": [
                    "versions.yml:md5,10fe5c201e5fcddb52c3607ab3fdfb34",
                    "versions.yml:md5,856745cef2fff087e50ea4c0ffa3addd",
                    "versions.yml:md5,a89255422a163684b0c80ebdd8ad28ae"
                ]
            }
        ],
        "timestamp": "2023-12-07T10:35:32.53584"
    }
}